[Bug]: No matching export in "global-externals:@storybook/channels" for import "createBrowserChannel"

### Describe the bug

I'm making the following dependency updates

```
-    "@storybook/addon-actions": "7.0.24",
+    "@storybook/addon-actions": "7.3.2",
-    "@storybook/addon-controls": "7.0.24",
+    "@storybook/addon-controls": "7.3.2",
-    "@storybook/addons": "7.0.24",
+    "@storybook/addons": "7.4.0",
-    "@storybook/theming": "7.0.24",
+    "@storybook/theming": "7.4.0",
```

and getting a failure on the `build-storybook` command:

```
Error: R] No matching export in "global-externals:@storybook/channels" for import "createBrowserChannel"
    ../../../../node_modules/@storybook/preview-api/dist/chunk-KAFJUYUA.mjs:5:9:
      5 │ import { createBrowserChannel } from '@storybook/channels';
        ╵          ~~~~~~~~~~~~~~~~~~~~
ERR! Error: Build failed with 1 error:
ERR! ../../../../node_modules/@storybook/preview-api/dist/chunk-KAFJUYUA.mjs:5:9: ERROR: No matching export in "global-externals:@storybook/channels" for import "createBrowserChannel"
```

Thanks for any help!

### To Reproduce

_No response_

### System

_No response_

### Additional context

_No response_

> Hey everyone! thanks for opening this issue. Can you provide a reproduction repo? I'd really like to get to the bottom of this issue and fix it.
> 
> Can you also check whether `npx storybook@latest automigrate` mentions any warnings like so in your project?
> 
> <img alt="image" width="644" src="https://user-images.githubusercontent.com/1671563/266035930-a4d2a802-f285-473d-9d0a-b295a2c6cdcd.png">

In our case for some reason we had this dependency in the project:
"storybook": "future"

It was pulling @storybook/cli 7.1.0-alpha.29 instead of 7.4.6, like the rest of the project. This caused the 7.1.0-alpha.29 to be used to compile storybooks generated using the newer versions. And as a result we got this error:

```
@storybook/cli v7.1.0-alpha.29

info => Cleaning outputDir: /build
info => Loading presets
info => Building manager..
✘ [ERROR] No matching export in "global-externals:@storybook/channels" for import "createBrowserChannel"

    ../../node_modules/@storybook/preview-api/dist/chunk-KAFJUYUA.mjs:5:9:
      5 │ import { createBrowserChannel } from '@storybook/channels';
        ╵          ~~~~~~~~~~~~~~~~~~~~

```

All it took was to remove "storybook" dependency. I guess, like noted in other comments, the issue is caused by either compiling or recompiling storybook using incompatible versions.

--- a/code/lib/core-events/src/errors/server-errors.ts
+++ b/code/lib/core-events/src/errors/server-errors.ts
@@ -385,3 +385,28 @@ export class NextjsSWCNotSupportedError extends StorybookError {
     `;
   }
 }
+
+export class NoMatchingExportError extends StorybookError {
+  readonly category = Category.CORE_SERVER;
+
+  readonly code = 4;
+
+  constructor(public data: { error: unknown | Error }) {
+    super();
+  }
+
+  template() {
+    return dedent`
+      There was an exports mismatch error when trying to build Storybook.
+      Please check whether the versions of your Storybook packages match whenever possible, as this might be the cause.
+      
+      Problematic example:
+      { "@storybook/react": "7.5.3", "@storybook/react-vite": "7.4.5", "storybook": "7.3.0" }
+
+      Correct example:
+      { "@storybook/react": "7.5.3", "@storybook/react-vite": "7.5.3", "storybook": "7.5.3" }
+
+      Clearing your lock file and reinstalling your dependencies might help as well.
+    `;
+  }
+}

--- a/code/lib/core-events/src/errors/server-errors.ts
+++ b/code/lib/core-events/src/errors/server-errors.ts
@@ -406,7 +406,7 @@ export class NoMatchingExportError extends StorybookError {
       Correct example:
       { "@storybook/react": "7.5.3", "@storybook/react-vite": "7.5.3", "storybook": "7.5.3" }
 
-      Clearing your lock file and reinstalling your dependencies might help as well.
+      Clearing your lock file and reinstalling your dependencies might help as well, as sometimes the version you see in your package.json might not be the one defined in your lock file, leading to version inconsistency issues.
     `;
   }
 }

--- a/code/lib/core-server/src/build-dev.ts
+++ b/code/lib/core-server/src/build-dev.ts
@@ -27,6 +27,7 @@ import { updateCheck } from './utils/update-check';
 import { getServerPort, getServerChannelUrl } from './utils/server-address';
 import { getManagerBuilder, getPreviewBuilder } from './utils/get-builders';
 import { warnOnIncompatibleAddons } from './utils/warnOnIncompatibleAddons';
+import { buildOrThrow } from './utils/build-or-throw';
 
 export async function buildDevStandalone(
   options: CLIOptions & LoadOptions & BuilderOptions
@@ -134,8 +135,8 @@ export async function buildDevStandalone(
     features,
   };
 
-  const { address, networkAddress, managerResult, previewResult } = await storybookDevServer(
-    fullOptions
+  const { address, networkAddress, managerResult, previewResult } = await buildOrThrow(async () =>
+    storybookDevServer(fullOptions)
   );
 
   const previewTotalTime = previewResult?.totalTime;

--- a/code/lib/core-server/src/build-static.ts
+++ b/code/lib/core-server/src/build-static.ts
@@ -35,6 +35,7 @@ import { extractStorybookMetadata } from './utils/metadata';
 import { StoryIndexGenerator } from './utils/StoryIndexGenerator';
 import { summarizeIndex } from './utils/summarizeIndex';
 import { defaultStaticDirs } from './utils/constants';
+import { buildOrThrow } from './utils/build-or-throw';
 
 export type BuildStaticStandaloneOptions = CLIOptions &
   LoadOptions &
@@ -146,7 +147,9 @@ export async function buildStaticStandalone(options: BuildStaticStandaloneOption
 
   global.FEATURES = features;
 
-  await managerBuilder.build({ startTime: process.hrtime(), options: fullOptions });
+  await buildOrThrow(async () =>
+    managerBuilder.build({ startTime: process.hrtime(), options: fullOptions })
+  );
 
   if (staticDirs) {
     effects.push(

--- a/code/lib/core-server/src/utils/build-or-throw.ts
+++ b/code/lib/core-server/src/utils/build-or-throw.ts
@@ -1,0 +1,20 @@
+import { NoMatchingExportError } from '@storybook/core-events/server-errors';
+
+export async function buildOrThrow<T>(callback: () => Promise<T>): Promise<T> {
+  try {
+    return await callback();
+  } catch (err: any) {
+    const builderErrors = err.errors as { text: string }[];
+    if (builderErrors) {
+      const inconsistentVersionsError = builderErrors.find((er) =>
+        er.text.includes('No matching export')
+      );
+
+      if (inconsistentVersionsError) {
+        throw new NoMatchingExportError(err);
+      }
+    }
+
+    throw err;
+  }
+}

--- a/code/lib/core-server/src/utils/build-or-throw.ts
+++ b/code/lib/core-server/src/utils/build-or-throw.ts
@@ -7,7 +7,7 @@ export async function buildOrThrow<T>(callback: () => Promise<T>): Promise<T> {
     const builderErrors = err.errors as { text: string }[];
     if (builderErrors) {
       const inconsistentVersionsError = builderErrors.find((er) =>
-        er.text.includes('No matching export')
+        er.text?.includes('No matching export')
       );
 
       if (inconsistentVersionsError) {

